Added tenant creation

rlvrs opened the pull request.

rlvrs committed:
Added tenant creation

Colex commented on spring-kotlin/aws-ssh-server-manager/ci/bootstrap_test_db.sh:
```diff
@@ -0,0 +1,11 @@
+    ALTER DEFAULT PRIVILEGES IN SCHEMA public GRANT SELECT, INSERT, UPDATE, DELETE ON TABLES TO myuser;
+    GRANT USAGE ON ALL SEQUENCES IN SCHEMA public TO myuser;
+    ALTER DEFAULT PRIVILEGES IN SCHEMA public GRANT USAGE ON SEQUENCES TO myuser;
+EOSQL
```
Nit: missing new line ;)

Colex replied:
what's this file?

rlvrs replied:
Oh, you're one of those :) 

This file just bootstraps the database that is spun up by my `docker-compose.yaml`
A common pattern we use that it is handy to check results
It might be also handy to perform some integration/system testing (calling the API and check results) later on

Colex commented on spring-kotlin/aws-ssh-server-manager/docker-compose.yaml:
```diff
@@ -0,0 +1,42 @@
+    image: adminer:4.7.7-standalone
+    restart: always
+    ports:
+      - "8080:8080"
```
Ditto

Colex commented on spring-kotlin/aws-ssh-server-manager/src/main/kotlin/dev/santos/awssshservermanager/config/WebSecurityConfig.kt:
```diff
@@ -0,0 +1,54 @@
+    override fun configure(auth: AuthenticationManagerBuilder) {
+        auth.userDetailsService(userDetailsService).passwordEncoder(passwordEncoder)
+    }
+}
```
does kotlin have a linter? It's usual for editors to auto-add a new line at the end

Colex replied:
also the spacing in this file looks quite odd

rlvrs replied:
Good shout: it should have! Will try [this one](https://ktlint.github.io/)

Colex commented on spring-kotlin/aws-ssh-server-manager/src/main/kotlin/dev/santos/awssshservermanager/config/WebSecurityConfig.kt:
```diff
@@ -0,0 +1,54 @@
+                        "/img/**",
+                        "/css/**",
+                        "/favicon.ico").permitAll()
+                .antMatchers("/api/v1/user/signup", "/api/v1/user/login").permitAll()
+                .antMatchers("/api/**").authenticated()
+                //.anyRequest().authenticated()
+                .and()
+                .formLogin()
+                .loginPage("/login.html")
+                .defaultSuccessUrl("/", true)
```
What is this repsponsible for? It's a bit worrying that it has hardcoded endpoints

rlvrs replied:
I didn't want to include this file so early on, but Controller tests were failing with 403, because Spring Security blocks everything by default

Authentication and Authorisation will be configured through this configuration

It is very likely way more verbose than it really needs

I can simplify it for now as this will be only important when I am validating FE->BE comms. I was planning on tweaking it once I started FE work.

Colex commented on spring-kotlin/aws-ssh-server-manager/src/main/kotlin/dev/santos/awssshservermanager/dto/CreateTenantDto.kt:
```diff
@@ -0,0 +1,21 @@
+      awsApiSecretSsmName = "${this.name}-api-secret"
+    )
+  }
+}
```
this one is good 😛

Colex commented on spring-kotlin/aws-ssh-server-manager/src/main/kotlin/dev/santos/awssshservermanager/dto/CreateTenantDto.kt:
```diff
@@ -0,0 +1,21 @@
+  val awsApiKey: String="",
+  val awsApiSecret: String=""
+) {
+  fun toTenant(): Tenant {
+    return Tenant(
+      name = this.name,
+      awsApiKeySsmName = "${this.name}-api-key",
+      awsApiSecretSsmName = "${this.name}-api-secret"
+    )
```
Is it a good practice to have logic within the DTO? I think that's the job of a service that will get the `CreateTenantDto` and create the Tenant. AFAIK the DTO just holds the data between the interfaces.

rlvrs replied:
Short answer: it is not. Thinking about it, Mapper would be more suitable!
Nice catch :eyes: :)

Colex commented on spring-kotlin/aws-ssh-server-manager/src/main/kotlin/dev/santos/awssshservermanager/lib/aws/config/AwsSsmConfig.kt:
```diff
@@ -0,0 +1,44 @@
+@Configuration
+class AwsSsmConfig {
+    @Value("\${aws.ssm.region}")
+    private val regionStr: String? = null
+
+    @Value("\${aws.ssm.endpointUri}")
+    private val endpointUriStr: String? = null
```
Is it normal in kotlin to `Str` at the end of the string? (The fact that their type is "string" should give that up)

rlvrs replied:
That is just me... because that endpointUri will be an URI later on and giving it the same name might be misleading
I actually don't need to assign it to a variable, so I can be less verbose as you suggest

I don't like to rely on scope to understand what the variable type is

Colex commented on spring-kotlin/aws-ssh-server-manager/src/main/kotlin/dev/santos/awssshservermanager/lib/aws/config/AwsSsmConfig.kt:
```diff
@@ -0,0 +1,44 @@
+    fun ssmClient(): SsmClient {
+        val region = Region.of(regionStr)
+        return when (endpointUriStr) {
+            null -> SsmClient
+                        .builder()
+                        .region(region)
+                        .build()
+            else -> SsmClient
+                        .builder()
+                        .region(region)
+                        .endpointOverride(URI(endpointUriStr))
+                        .build()
```
How do you test these chained things? Seems to be a java pattern, but I don't understand the advantage over the patterns used in typescript and ruby:
```typescript
const client = new SsmClient({ region: region, endpointOverride: endpointUriStr });
```

Colex replied:
It looks like it's forcing to do some code duplication as well

rlvrs replied:
I can make it DRYer for sure, but there is no way to do it as you suggest (python is as flexible as typescript and ruby, btw)

That is the kind of thing I just test against an environment or docker-compose generated environment

IMHO, though it would be possible, there is no need to do that at the unit level as it provides little confidence

rlvrs replied:
I'm going to delete the code, but it would look something like this:
```
  fun ssmClient(): SsmClient {
    val builder = SsmClient.builder()

    if (!endpointUri.isNullOrBlank()) {
      builder.endpointOverride(URI(endpointUri))
    }

    return builder.region(Region.of(region))
      .build()
  }
```

rlvrs replied:
Also, this is called builder pattern. Quite common in the JVM/typed languages world
That dictionary of params (kwargs in Python) has not been a common practice
Though it can be achieved with a `data class` as a parameter. But data classes are recent enough and the builder pattern is widely spread atm

Colex commented on spring-kotlin/aws-ssh-server-manager/src/main/kotlin/dev/santos/awssshservermanager/model/Tenant.kt:
```diff
@@ -0,0 +1,14 @@
+  @GeneratedValue(strategy = GenerationType.IDENTITY)
+  val id:Long=0,
+  val name: String = "",
+  val awsApiKeySsmName: String = "",
+  val awsApiSecretSsmName: String = ""
```
Why does it have names? It should be just `awsAccessKeyId` and `awsSecretAccessKey`

rlvrs replied:
As agreed, I'm going to store this in the DB directly encrypt it at rest
Thus, I'll rename this as you suggest

rlvrs committed:
Addressed tenant creation PR comments:
 - Added linter
 - Simplified WebSecurityConfig
 - Removed SSM Parameter store dependency
 - Added Mapper to convert DTO to Model

rlvrs committed:
Improvements based on Tenant creation PR comments:
 - Remove SSM dependencies and configs

migueltaoliveira commented on spring-kotlin/aws-ssh-server-manager/src/main/kotlin/dev/santos/awssshservermanager/mapper/TenantMapper.kt:
```diff
@@ -0,0 +1,23 @@
+)
+interface TenantMapper {
+  @Mappings(
+    Mapping(source = "name", target = "name"),
```
You can omit mapping when source and target are the same.

migueltaoliveira commented on spring-kotlin/aws-ssh-server-manager/src/main/kotlin/dev/santos/awssshservermanager/service/TenantService.kt:
```diff
@@ -0,0 +1,30 @@
+      return tenantRepository
+        .save(newTenant)
+        .id
+    } catch (exception: Exception) {
```
Why don't you only catch the DataIntegrityViolationException instead? The rest of the exceptions will be thrown anyway.

migueltaoliveira commented on spring-kotlin/aws-ssh-server-manager/src/main/kotlin/dev/santos/awssshservermanager/dto/CreateTenantDto.kt:
```diff
@@ -0,0 +1,12 @@
+  @field:Pattern(regexp = "^[a-z]+(-[a-z\\d]+)*\$")
+  @field:Size(min = 2, max = 50)
+  val name: String = "",
+  val awsApiKey: String = "",
```
You should limit the size of each one of the fields to 250 since you allocated 250 characters in the database.

rlvrs commented on spring-kotlin/aws-ssh-server-manager/src/main/kotlin/dev/santos/awssshservermanager/config/WebSecurityConfig.kt:
```diff
@@ -0,0 +1,15 @@
+import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity
+import org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter
+
+@Configuration
```
Use test configuration

rlvrs commented on spring-kotlin/aws-ssh-server-manager/src/main/kotlin/dev/santos/awssshservermanager/helper/JsonHelper.kt:
```diff
@@ -0,0 +1,7 @@
+
+fun objToJsonStr(obj: Any): String = Gson().toJson(obj)
+
+inline fun <reified T> strToJsonObj(objStr: String): T = Gson().fromJson(objStr, T::class.java)
```
new line

rlvrs commented on spring-kotlin/aws-ssh-server-manager/src/test/kotlin/dev/santos/awssshservermanager/controller/TenantControllerShould.kt:
```diff
@@ -0,0 +1,106 @@
+      accept = MediaType.APPLICATION_JSON
+    }.andExpect {
+      status { isCreated() }
+      content { MockMvcResultMatchers.status().isBadRequest }
```
remoove

rlvrs commented on spring-kotlin/aws-ssh-server-manager/src/test/kotlin/dev/santos/awssshservermanager/controller/TenantControllerShould.kt:
```diff
@@ -0,0 +1,106 @@
+      status { isCreated() }
+      content { MockMvcResultMatchers.status().isBadRequest }
+      content { contentType(MediaType.APPLICATION_JSON) }
+      content { json("{\"id\":1}") }
```
json serialization

rlvrs commented on spring-kotlin/aws-ssh-server-manager/src/main/kotlin/dev/santos/awssshservermanager/controller/TenantController.kt:
```diff
@@ -0,0 +1,20 @@
+  fun createTenant(@Valid @RequestBody createTenantDto: CreateTenantDto): ResponseEntity<String> {
+    val newTenantId = tenantService.create(createTenantDto)
+    val location: URI = URI.create("/tenant/${createTenantDto.name}")
+    return ResponseEntity.created(location).body("{\"id\":${newTenantId}}")
```
json serialization

rlvrs commented on spring-kotlin/aws-ssh-server-manager/src/test/kotlin/dev/santos/awssshservermanager/controller/TenantControllerShould.kt:
```diff
@@ -0,0 +1,106 @@
+      accept = MediaType.APPLICATION_JSON
+    }.andExpect {
+      status { isConflict() }
+    }
```
missing fail

rlvrs commented on spring-kotlin/aws-ssh-server-manager/src/test/kotlin/dev/santos/awssshservermanager/controller/TenantControllerShould.kt:
```diff
@@ -0,0 +1,106 @@
+      awsApiSecret = "super_secret_secret"
+    )
+    BDDMockito.given(this.tenantService.create(testTenantDto))
+      .willAnswer {
```
willThrow

rlvrs replied:
investigate checked VS unchecked exceptions (suppa low priority)

rlvrs commented on spring-kotlin/aws-ssh-server-manager/src/test/kotlin/dev/santos/awssshservermanager/helper/TestMatchersHelpers.kt:
```diff
@@ -0,0 +1,5 @@
+
+import org.mockito.ArgumentMatchers
+
+fun <T : Any> safeEq(value: T): T = ArgumentMatchers.eq(value) ?: value
```
new line

rlvrs commented on spring-kotlin/aws-ssh-server-manager/src/test/resources/mockito-extensions/org.mockito.plugins.MockMaker:
```diff
@@ -0,0 +1 @@
+mock-maker-inline
```
new line

rlvrs committed:
Addressed PR comments:
 - API Response serialization
 - Nullable Request DTO to be easier to test
 - Removed unnecessary mapstruct mappings
 - Properly test check exception catch
 - Do not Mock mapper
